prow/crier/reporters/gcs/reporter.go

```diff
@@ -74,24 +75,60 @@ func (gr *gcsReporter) reportJobState(ctx context.Context, log *logrus.Entry, pj
 // happen before the pod itself gets to upload one, at which point the pod will
 // upload its own. If for some reason one already exists, it will not be overwritten.
 func (gr *gcsReporter) reportStartedJob(ctx context.Context, log *logrus.Entry, pj *prowv1.ProwJob) error {
-	s := downwardapi.PjToStarted(pj, nil)
+	bucketName, dir, err := util.GetJobDestination(gr.cfg, pj)
+	if err != nil {
+		return fmt.Errorf("failed to get job destination: %w", err)
+	}
+
+	if gr.dryRun {
+		log.WithFields(logrus.Fields{"bucketName": bucketName, "dir": dir}).Debug("Would upload started.json")
+		return nil
+	}
+
+	// Try to read clone records
+	var cloneRecord []clone.Record
+	cloneRecordBytes, err := io.ReadContent(ctx, log, gr.opener, providers.GCSStoragePath(bucketName, path.Join(dir, prowv1.CloneRecordFile)))
+	if err != nil {
+		if !io.IsNotExist(err) {
+			log.WithError(err).Warn("Failed to read clone records.")
+		}
+	} else {
+		if err := json.Unmarshal(cloneRecordBytes, &cloneRecord); err != nil {
+			log.WithError(err).Warn("Failed to unmarshal clone records.")
+		}
+	}
+	s := downwardapi.PjToStarted(pj, cloneRecord)
 	s.Metadata = metadata.Metadata{"uploader": "crier"}
 
 	output, err := json.MarshalIndent(s, "", "\t")
 	if err != nil {
 		return fmt.Errorf("failed to marshal started metadata: %w", err)
 	}
 
-	bucketName, dir, err := util.GetJobDestination(gr.cfg, pj)
-	if err != nil {
-		return fmt.Errorf("failed to get job destination: %w", err)
+	// Best-effort read of existing started.json; it's overwritten only if it's uploaded
+	// by crier and there is something new (clone record).
+	var existingStarted metadata.Started
+	content, err := io.ReadContent(ctx, log, gr.opener, providers.GCSStoragePath(bucketName, path.Join(dir, prowv1.StartedStatusFile)))
+	if err == nil {
+		if err = json.Unmarshal(content, &existingStarted); err != nil {
+			log.WithError(err).Warn("Failed to unmarshal started.json.")
+		}
+	} else {
+		if !io.IsNotExist(err) {
+			log.WithError(err).Warn("Failed to read started.json.")
+		}
 	}
 
-	if gr.dryRun {
-		log.WithFields(logrus.Fields{"bucketName": bucketName, "dir": dir}).Debug("Would upload started.json")
+	// Overwrite if it was uploaded by crier and there might be something new
+	var opts []io.WriterOptions
+	if existingStarted.Metadata["uploader"] == "crier" && existingStarted.RepoCommit != s.RepoCommit {
+		opts = append(opts, io.WriterOptions{PreconditionDoesNotExist: utilpointer.BoolPtr(false)})
+	}
+	// Skip upload if file already exists and there is nothing new
+	if err == nil && len(opts) == 0 {
```

Comment:
Can we do these checks before reading the clone record so that we avoid that extra GCS read when we see an existing started.json with a populated repo commit?

Reply:
good point, done and added a unit test for it